packages/mui-material-next/src/FilledInput/FilledInput.tsx

```diff
@@ -145,115 +155,174 @@ const FilledInputInput = styled(InputBaseInput, {
+  RootComponentType extends React.ElementType,
+>(inProps: FilledInputProps<RootComponentType>, forwardedRef: React.ForwardedRef<Element>) {
   const props = useThemeProps({ props: inProps, name: 'MuiFilledInput' });
 
   const {
     disableUnderline,
-    components = {},
-    componentsProps: componentsPropsProp,
     fullWidth = false,
     hiddenLabel, // declare here to prevent spreading to DOM
     inputComponent = 'input',
     multiline = false,
-    slotProps,
-    slots = {},
     type = 'text',
+    slotProps = {},
+    slots = {},
     ...other
   } = props;
 
-  const ownerState = {
+  const ownerState: FilledInputOwnerState = {
     ...props,
+    disableUnderline,
     fullWidth,
     inputComponent,
     multiline,
     type,
   };
 
-  const classes = useUtilityClasses(props);
-  const filledInputComponentsProps = { root: { ownerState }, input: { ownerState } };
+  const classes = useUtilityClasses(ownerState);
+
+  const Root = slots.root ?? FilledInputRoot;
+  const Input = slots.input ?? FilledInputInput;
+
+  const rootProps = useSlotProps({
+    elementType: Root,
+    externalSlotProps: slotProps.root,
+    additionalProps: {
+      ref: forwardedRef,
+      fullWidth,
+      inputComponent,
+    },
+    externalForwardedProps: other,
+    ownerState: ownerState as FilledInputOwnerState & InputBaseOwnerState,
+    className: [classes.root],
+  });
 
-  const componentsProps =
-    slotProps ?? componentsPropsProp
-      ? deepmerge(slotProps ?? componentsPropsProp, filledInputComponentsProps)
-      : filledInputComponentsProps;
+  const inputProps = useSlotProps({
+    elementType: Input,
+    externalSlotProps: slotProps.input,
+    ownerState: ownerState as FilledInputOwnerState & InputBaseOwnerState,
+    className: [classes.input],
+  });
 
-  const RootSlot = slots.root ?? components.Root ?? FilledInputRoot;
-  const InputSlot = slots.input ?? components.Input ?? FilledInputInput;
+  if (multiline) {
+    return (
+      <InputBase
+        slots={{ root: Root, textarea: Input }}
+        slotProps={{
+          input: inputProps,
+        }}
+        multiline
+        {...rootProps}
+      />
+    );
+  }
 
   return (
     <InputBase
-      slots={{ root: RootSlot, input: InputSlot }}
-      componentsProps={componentsProps}
-      fullWidth={fullWidth}
-      inputComponent={inputComponent}
-      multiline={multiline}
-      ref={ref}
+      slots={{ root: Root, input: Input }}
+      slotProps={{
+        input: inputProps,
+      }}
       type={type}
-      {...other}
-      classes={classes}
+      {...rootProps}
     />
   );
-});
+}) as FilledInputComponent;
+
+interface FilledInputComponent {
+  <C extends React.ElementType>(
+    props: {
+      /**
+       * The component used for the input node.
+       * Either a string to use a HTML element or a component.
+       * @default 'input'
+       */
+      inputComponent?: C;
+    } & OverrideProps<FilledInputTypeMap, C>,
+  ): JSX.Element | null;
```

Comment:
Isn't this missing a declaration for the default component type as: https://github.com/mui/material-ui/blob/v5.14.13/packages/mui-types/index.d.ts#L94

Also, could we use `OverridableComponent`?

Reply:
Updated!

I didn't use `OverridableComponent` because like `InputBase`, the `component` prop is called `inputComponent` instead

Reply:
> Don't we want to standardize the use of `slots` and `slotProps` in v6?

Yes! But here the `inputComponent` prop is the one that gets passed to emotion's `as`, so no conflict with `slots`

(The naming is confusing, confusing enough I [accidentally removed this](https://github.com/mui/material-ui/pull/39032#issuecomment-1737480052) before 🫠)

Reply:
Ah, so it's an equivalent of `slotProps.input.component` I wonder if it's used so frequently it deserves being lifted to the owner component, but that's a discussion for later, when we'll clean the API.